### ISSUE TITLE
Simply theme API

### ISSUE THREAD

**Issue by andydotxyz** [fyne-io/fyne#4141]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

In discussion of #4040 we realised that the theme package API is getting larger all the time.
The v2 `Theme` type was introduced to make theme extensions backward compatible, but each new theme name adds more functions to the package.

### Is it possible to construct a solution with the existing API?

Yes, but it is getting bulky

### Describe the solution you'd like to see.

Proposal is we should move to fewer functions over time - theme.PrimaryColor() plus the many others could be `theme.CurrentColor(ColorName)` or theme.`ColorNamed(ColorName)` or similar to avoid this api explosion...

In addition we could consider deprecating the colour name specific themed resource functions and types in preference of `theme.NewColoredResource` added in v2.4.0

**Comment by pseidemann** [comment]
hi,
what do you think about using a simple struct (with some more nested structs for colors etc.)? zero values will mean that the default value will be used.

**Comment by andydotxyz** [comment]
> hi,
> 
> what do you think about using a simple struct (with some more nested structs for colors etc.)? zero values will mean that the default value will be used.
> 


Using concrete structs makes it much harder to extend. What be edit do you think this conveys over the interface we have currently? (fyne.Theme)

**Comment by pseidemann** [comment]
according to official go docs, a struct is actually more compatible when doing changes than an interface, see https://go.dev/blog/module-compatibility#maintaining-struct-compatibility.

they give the example of the ListenConfig which is a struct: https://pkg.go.dev/net@go1.13#ListenConfig

**Comment by andydotxyz** [comment]
That is true if you are talking about adding functionality to a type you own. In our case neither applies, we are adding new possible options to existing functions on types that are controlled by other developers. Interface is perfect here as we are not changing the contract and don't control the code.

**Comment by pseidemann** [comment]
maybe I'm missing something but I'm not sure what you mean by "own". the interface `fyne.Theme` is defined by fyne and can be implemented by fyne consumers. fyne also happens to provide a default implementation of that interface.

now if fyne adds a new function to that theme interface, all existing consumer projects will be broken, since they will all miss that new function.

if a struct would be in use, these projects would not break and the field will just become the zero value, defaulting to the default theme implementation.

I acknowledge that there is a special case, were just the allowed arguments for the *already existing* functions are changed, without changing the function's signature. I guess this works, since strings are used (and enums aren't used / go doesn't support proper enums which actually restrict passed values), so we have basically infinite different values we could pass in the future. I can see that this should be mostly backwards compatible, since existing implementations will just not change newly introduced values and the default theme will be applied again, if the implementation correctly falls back to looking up the default value from the default theme. if the implementation would fall back to some other (hardcoded) value, then it would break again those projects.

**Comment by andydotxyz** [comment]
Oh I understand - you would propose a single fyne provided struct where every possible theme attribute is a field?

Gosh... you realise how huge that would be?

* 7 fonts
* 90 icons
* 14 sizes
* 23 colours (in 2 variants - currently) = 46

So that struct would have 157 fields in it...

I'm not sure that sounds like a desirable API to be able to handle the case where themes in the future need a new type of attribute that our current API has not planned for.

**Comment by Jacalz** [comment]
Ignoring the fact that a global variable holding a struct wouldn't be immutable and slow to access compared to APIs that can be inlined into code, the memory usage of that struct would be gigantic.

**Comment by pseidemann** [comment]
of course you would break down that struct into smaller nested structs as per my initial message. for icons maybe something else could/should be used. a map? but this is just a suggestion. if you think this isn't feasible or makes sense then I'm ok with that 👍 

> Ignoring the fact that a global variable holding a struct wouldn't be immutable and slow to access compared to APIs that can be inlined into code, the memory usage of that struct would be gigantic.

just curious: why would it be slower to access?

**Comment by andydotxyz** [comment]
> of course you would break down that struct 

There's not much of an "of course" about it. If you want to propose an alternative API design we will evaluate it based on what you put forward.

I think we can conclude at this time there is not such a compelling case to make a breaking change to move to this alternative format in an upcoming release.
